File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/main/MainActivityPresenter.java
package app.editors.manager.mvp.presenters.main;

import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.android.play.core.review.ReviewInfo;
import com.google.android.play.core.review.ReviewManagerFactory;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import app.editors.manager.BuildConfig;
import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.app.WebDavApi;
import app.editors.manager.managers.exceptions.UrlSyntaxMistake;
import app.editors.manager.managers.utils.FirebaseUtils;
import app.editors.manager.managers.utils.GlideUtils;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.mvp.models.response.ResponseUser;
import app.editors.manager.mvp.models.user.User;
import app.editors.manager.mvp.presenters.base.BasePresenter;
import app.editors.manager.mvp.views.main.MainActivityView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import lib.toolkit.base.managers.tools.GlideTool;
import lib.toolkit.base.managers.utils.NetworkUtils;
import lib.toolkit.base.managers.utils.StringUtils;
import moxy.InjectViewState;
import retrofit2.HttpException;

@InjectViewState
public class MainActivityPresenter extends BasePresenter<MainActivityView, ResponseUser>
        implements FirebaseUtils.OnRatingApp {

    public static final String TAG = MainActivityPresenter.class.getSimpleName();

    public static final String TAG_DIALOG_REMOTE_PLAY_MARKET = "TAG_DIALOG_REMOTE_PLAY_MARKET";
    public static final String TAG_DIALOG_REMOTE_APP = "TAG_DIALOG_REMOTE_APP";
    public static final String TAG_DIALOG_RATE_FIRST = "TAG_DIALOG_RATE_FIRST";
    private static final String TAG_DIALOG_RATE_SECOND = "TAG_DIALOG_RATE_SECOND";
    private static final String TAG_DIALOG_RATE_FEEDBACK = "TAG_DIALOG_RATE_FEEDBACK";

    private static final long DEFAULT_RATE_SESSIONS = 5;

    private static boolean sIsAppColdStart = true;
    private AccountsSqlData mAccountsSqlData;
    private CompositeDisposable mDisposable = new CompositeDisposable();

    @Inject
    GlideTool mGlideTool;

    @Nullable
    private AccountsSqlData mAccount;
    private boolean mRestoreFlag = false;
    private boolean mIsDialogOpen = false;

    @Nullable
    private ReviewInfo mReviewInfo;

    public MainActivityPresenter() {
        App.getApp().getAppComponent().inject(this);
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        mPreferenceTool.setUserSession();
        if (sIsAppColdStart) {
            sIsAppColdStart = false;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mDisposable.dispose();
    }

    @Override
    public void onRatingApp(boolean isRating) {
        if (isRating) {
            if (mPreferenceTool.getIsRateOn() && (mPreferenceTool.getUserSession() % DEFAULT_RATE_SESSIONS) == 0) {
                getViewState().onRatingApp();
            }
        }
    }

    public void getRemoteConfigRate() {
        if (!BuildConfig.DEBUG) {
            mDisposable.add(Observable.just(1)
                    .delay(500, TimeUnit.MILLISECONDS)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(integer -> FirebaseUtils.checkRatingConfig(MainActivityPresenter.this)));
        }
    }

    public void onAcceptClick(@Nullable String value, @Nullable String tag) {
        if (tag != null) {
            switch (tag) {
                case TAG_DIALOG_REMOTE_PLAY_MARKET:
                    getViewState().onShowPlayMarket(BuildConfig.RELEASE_ID);
                    getViewState().onDialogClose();
                    break;
                case TAG_DIALOG_REMOTE_APP:
                    getViewState().onShowApp(BuildConfig.RELEASE_ID);
                    getViewState().onDialogClose();
                    break;
                case TAG_DIALOG_RATE_FIRST:
                    getReviewInfo();
                    getViewState().onQuestionDialog(mContext.getString(R.string.dialogs_question_rate_second_info), TAG_DIALOG_RATE_SECOND,
                            mContext.getString(R.string.dialogs_question_accept_sure),
                            mContext.getString(R.string.dialogs_question_accept_no_thanks), null);
                    break;
                case TAG_DIALOG_RATE_SECOND:
                    mPreferenceTool.setIsRateOn(false);
                    if (mReviewInfo != null) {
                        getViewState().onShowInAppReview(mReviewInfo);
                    } else {
                        getViewState().onShowPlayMarket(BuildConfig.RELEASE_ID);
                    }
                    getViewState().onDialogClose();
                    break;
                case TAG_DIALOG_RATE_FEEDBACK:
                    if (value != null) {
                        getViewState().onShowEmailClientTemplate(value);
                    }
                    getViewState().onDialogClose();
                    break;
            }
        }
    }

    private void getReviewInfo() {
        ReviewManagerFactory.create(mContext).requestReviewFlow().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                mReviewInfo = task.getResult();
            }
        });
    }

    public void onCancelClick(String tag) {
        if (tag != null) {
            switch (tag) {
                case TAG_DIALOG_RATE_FIRST:
                    mPreferenceTool.setIsRateOn(false);
                    getViewState().onShowEditMultilineDialog(mContext.getString(R.string.dialogs_edit_feedback_rate_title),
                            mContext.getString(R.string.dialogs_edit_feedback_rate_hint),
                            mContext.getString(R.string.dialogs_edit_feedback_rate_accept),
                            mContext.getString(R.string.dialogs_question_accept_no_thanks), TAG_DIALOG_RATE_FEEDBACK);
                    return;
                case TAG_DIALOG_RATE_SECOND:
                    mPreferenceTool.setIsRateOn(false);
                    getViewState().onDialogClose();
                    break;
            }
        }
    }

    public void checkOnBoarding() {
        if (!mPreferenceTool.getOnBoarding()) {
            getViewState().onShowOnBoarding();
        }
    }

    public void navigationItemClick(int itemId) {

        getViewState().onCloseActionDialog();

        if (!mRestoreFlag) {
            getViewState().onClearStack();
        }
        switch (itemId) {
            case R.id.menu_item_recent:
                getViewState().onShowRecentFragment(mRestoreFlag);
                break;
            case R.id.menu_item_cloud:
                if (mAccount != null && !mPreferenceTool.isNoPortal()) {
                    if (mAccount.isWebDav()) {
                        getViewState().onShowWebDavFragment(mRestoreFlag, WebDavApi.Providers.valueOf(mAccount.getWebDavProvider()));
                    } else {
                        getViewState().onShowCloudFragment(mRestoreFlag, false);
                    }
                } else {
                    getViewState().onShowCloudFragment(mRestoreFlag, true);
                    cancelGetInfo();
                }
                break;
            case R.id.menu_item_on_device:
                getViewState().onShowOnDeviceFragment(mRestoreFlag);
                cancelGetInfo();
                break;
            case R.id.menu_item_setting:
                if (mAccountSqlTool.getAccounts() != null && !mAccountSqlTool.getAccounts().isEmpty()) {
                    getViewState().onShowAccountsFragment(mRestoreFlag);
                } else {
                    getViewState().onShowProfileFragment();
                }
                cancelGetInfo();
                break;
        }
        mRestoreFlag = false;
    }

    public void setAccount() {
        mAccount = mAccountSqlTool.getAccountOnline();
    }

    public void setToolbarAccount(boolean isVisible) {
        final AccountsSqlData account = mAccountSqlTool.getAccountOnline();
        if (account != null) {
            final String portal = account.getPortal();
            String login;
            if (account.isWebDav()) {
                login = account.getLogin();
                getViewState().onSetWebDavImage(WebDavApi.Providers.valueOf(account.getWebDavProvider()));
            } else {
                loadAvatar(account);
                login = account.getName();
            }
            getViewState().onShowToolbarAccount(portal, login, isVisible);
        }
    }

    public void setRestore(boolean isRestore) {
        mRestoreFlag = isRestore;
    }

    public void checkPortal() {
        if (mAccount != null && mAccount.isWebDav() && mAccount.getPassword() != null && !mAccount.getPassword().isEmpty()) {
            mPreferenceTool.setNoPortal(false);
        } else if (mPreferenceTool.getLogin() != null && mPreferenceTool.getToken() != null) {
            mPreferenceTool.setNoPortal(false);
        } else {
            mPreferenceTool.setNoPortal(true);
        }
    }

    public void setUser() {
        if (mAccountsSqlData != null) {
            mPreferenceTool.setPortal(mAccountsSqlData.getPortal());
            mPreferenceTool.setSslCiphers(mAccountsSqlData.isSslCiphers());
            mPreferenceTool.setSslState(mAccountsSqlData.isSslState());
            mPreferenceTool.setScheme(mAccountsSqlData.getScheme());
            mPreferenceTool.setNoPortal(false);
            mPreferenceTool.setLogin(mAccountsSqlData.getLogin());
            mPreferenceTool.setToken(mAccountsSqlData.getToken());
            mPreferenceTool.setPassword(mAccountsSqlData.getPassword());
            try {
                mRetrofitTool.setSslOn(mAccountsSqlData.isSslState());
                mRetrofitTool.setCiphers(mAccountsSqlData.isSslCiphers());
                mRetrofitTool.init(mAccountsSqlData.getScheme() + StringUtils.getEncodedString(mAccountsSqlData.getPortal()));
            } catch (UrlSyntaxMistake urlSyntaxMistake) {
                urlSyntaxMistake.printStackTrace();
            }
        }
        mAccountsSqlData = null;
    }

    public void getAccount() {
        if (mPreferenceTool.getPortal() != null) {
            mAccountsSqlData = mAccountSqlTool
                    .getAccount(mPreferenceTool.getPortal(), mPreferenceTool.getLogin(), mPreferenceTool.getSocialProvider());
        } else {
            mAccountsSqlData = null;
        }
    }

    public void clearAccount() {
        mPreferenceTool.setDefault();
        AccountsSqlData accountsSqlData = mAccountSqlTool.getAccountOnline();
        if (accountsSqlData != null) {
            accountsSqlData.setOnline(false);
            mAccountSqlTool.setAccount(accountsSqlData);
        }
        if (mAccount != null) {
            mAccount.setOnline(false);
            mAccountSqlTool.setAccount(mAccount);
        }
        mAccount = null;
    }

    public void checkAccountInfo() {
        if (mAccount != null && !mAccount.isWebDav()) {
            mDisposable.add(Observable.fromCallable(() -> mRetrofitTool.getApiWithPreferences().getUserInfo(mPreferenceTool.getToken()).execute())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .map(responseUserResponse -> {
                        if (responseUserResponse.isSuccessful() && responseUserResponse.body() != null) {
                            return responseUserResponse.body().getResponse();
                        } else {
                            throw new HttpException(responseUserResponse);
                        }
                    })
                    .filter(user -> !user.getEmail().equals(""))
                    .subscribe(this::setUserInfo, throwable -> {
                        if (throwable instanceof HttpException) {
                            final int code = ((HttpException) throwable).code();
                            if (code == Api.HttpCodes.CLIENT_UNAUTHORIZED) {
                                getViewState().onUnauthorized(mContext.getString(R.string.errors_client_unauthorized));
                            }
                        }
                    }));

        }
    }

    private void setUserInfo(User user) {
        mPreferenceTool.setAdmin(user.getIsAdmin());
        mPreferenceTool.setOwner(user.getIsOwner());
        mPreferenceTool.setVisitor(user.getIsVisitor());
        if (mAccount != null) {
            if (!mAccount.getName().equals(user.getDisplayNameHtml())) {
                mAccount.setName(user.getDisplayNameHtml());
            }
            if (!mAccount.getLogin().equals(user.getEmail())) {
                mAccount.setLogin(user.getEmail());
            }
            if (!mAccount.getAvatarUrl().equals(user.getAvatar())) {
                mAccount.setAvatarUrl(user.getAvatar());
                loadAvatar(mAccount);
            }
            mAccountSqlTool.setAccount(mAccount);
            loadAvatar(mAccount);
            //getViewState().onShowToolbarAccount(mAccount.getPortal(), mAccount.getName(), true);
        }
    }

    private void loadAvatar(AccountsSqlData account) {
        if (NetworkUtils.isOnline(mContext)) {
            Glide.with(mContext)
                    .load(GlideUtils.getCorrectLoad(account.getAvatarUrl(), mPreferenceTool))
                    .apply(GlideUtils.getAvatarOptions())
                    .into(getSimpleTarget());
        }
    }

    private CustomTarget<Drawable> getSimpleTarget() {
        return new CustomTarget<Drawable>() {

            @Override
            public void onResourceReady(@NonNull Drawable resource, Transition<? super Drawable> transition) {
                getViewState().onAccountAvatar(resource);
            }

            @Override
            public void onLoadCleared(@Nullable Drawable placeholder) {
                if (placeholder instanceof BitmapDrawable) {
                    getViewState().onAccountAvatar(placeholder);
                }

            }
        };
    }

    private void cancelGetInfo() {
        mDisposable.clear();
        getViewState().onShowToolbarAccount("", "", false);
    }

    public void setDialogOpen(boolean isOpen) {
        mIsDialogOpen = isOpen;
    }

    public boolean isDialogOpen() {
        return mIsDialogOpen;
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/main/MainActivityPresenter.kt
package app.editors.manager.mvp.presenters.main

import android.net.Uri
import android.os.Build
import androidx.core.net.toUri
import app.documents.core.account.AccountPreferences
import app.documents.core.account.AccountRepository
import app.documents.core.database.datasource.RecentDataSource
import app.documents.core.login.CheckLoginResult
import app.documents.core.login.PortalResult
import app.documents.core.model.cloud.CloudPortal
import app.documents.core.model.cloud.Recent
import app.documents.core.model.cloud.Scheme
import app.documents.core.model.cloud.isDocSpace
import app.documents.core.network.common.NetworkResult
import app.documents.core.providers.FileOpenResult
import app.editors.manager.BuildConfig
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.app.accountOnline
import app.editors.manager.app.appComponent
import app.editors.manager.app.cloudFileProvider
import app.editors.manager.mvp.models.filter.Filter
import app.editors.manager.mvp.models.models.OpenDataModel
import app.editors.manager.mvp.presenters.base.BasePresenter
import app.editors.manager.mvp.views.main.MainActivityView
import com.google.android.gms.tasks.Task
import com.google.android.play.core.review.ReviewInfo
import com.google.android.play.core.review.ReviewManagerFactory
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import lib.toolkit.base.managers.utils.CryptUtils
import lib.toolkit.base.managers.utils.FileUtils
import moxy.InjectViewState
import moxy.presenterScope
import javax.inject.Inject

@InjectViewState
class MainActivityPresenter : BasePresenter<MainActivityView>() {

    companion object {
        val TAG: String = MainActivityPresenter::class.java.simpleName
        const val TAG_DIALOG_REMOTE_PLAY_MARKET = "TAG_DIALOG_REMOTE_PLAY_MARKET"
        const val TAG_DIALOG_REMOTE_APP = "TAG_DIALOG_REMOTE_APP"
        const val TAG_DIALOG_RATE_FIRST = "TAG_DIALOG_RATE_FIRST"
        private const val TAG_DIALOG_RATE_SECOND = "TAG_DIALOG_RATE_SECOND"
        private const val TAG_DIALOG_RATE_FEEDBACK = "TAG_DIALOG_RATE_FEEDBACK"
        private const val DEFAULT_RATE_SESSIONS: Long = 5
    }

    init {
        App.getApp().appComponent.inject(this)
    }

    @Inject
    lateinit var accountPreferences: AccountPreferences

    @Inject
    lateinit var accountRepository: AccountRepository

    @Inject
    lateinit var recentDataSource: RecentDataSource

    private val disposable = CompositeDisposable()

    private var reviewInfo: ReviewInfo? = null
    var isAppColdStart = true
        private set

    var isDialogOpen: Boolean = false
    val isVPNChecked: Boolean
        get() = preferenceTool.isVpnChecked

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        preferenceTool.setUserSession()
        preferenceTool.filter = Filter()
        checkSdk()
    }

    private fun checkSdk() {
        presenterScope.launch(Dispatchers.IO) {
            val version = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                context.packageManager.getPackageInfo(context.packageName, 0).longVersionCode
            } else {
                @Suppress("DEPRECATION")
                context.packageManager.getPackageInfo(context.packageName, 0).versionCode.toLong()
            }

            if (preferenceTool.appVersion != BuildConfig.VERSION_NAME + "." + version) {
                context.externalCacheDir?.let(FileUtils::deletePath)
                context.cacheDir?.let(FileUtils::deletePath)
            }
            preferenceTool.appVersion = BuildConfig.VERSION_NAME + "." + version
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }

    fun init(isPortal: Boolean = false) {
        presenterScope.launch {
            context.accountOnline?.let { account ->
                if (isAppColdStart) {
                    App.getApp().refreshLoginComponent(account.portal)
                    if (account.isDocSpace) {
                        val isRegular = App.getApp().loginComponent.cloudLoginRepository.checkUserRegular()
                        accountPreferences.isRegularUser = isRegular
                    } else {
                        accountPreferences.isRegularUser = false
                    }
                    App.getApp().loginComponent.cloudLoginRepository.updatePortalSettings()
                    isAppColdStart = false
                }
            }
        }
        checkAppLocale()
    }

    private fun checkAppLocale() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (context.appComponent.appLocaleHelper.checkAppLocale(true)) {
                viewState.onLocaleConfirmation()
            }
        }
    }

    fun getRemoteConfigRate() {
        if (!BuildConfig.DEBUG) {
            if (preferenceTool.isRateOn && preferenceTool.userSession % DEFAULT_RATE_SESSIONS == 0L) {
                viewState.onRatingApp()
            }
        }
    }

    private fun getReviewInfo() {
        ReviewManagerFactory.create(context)
            .requestReviewFlow()
            .addOnCompleteListener { task: Task<ReviewInfo?> ->
                if (task.isSuccessful) {
                    reviewInfo = task.result
                }
            }
    }

    fun onAcceptClick(value: String?, tag: String?) {
        tag?.let {
            when (tag) {
                TAG_DIALOG_REMOTE_PLAY_MARKET -> {
                    viewState.onShowPlayMarket(BuildConfig.RELEASE_ID)
                    viewState.onDialogClose()
                }

                TAG_DIALOG_REMOTE_APP -> {
                    viewState.onShowApp(BuildConfig.RELEASE_ID)
                    viewState.onDialogClose()
                }

                TAG_DIALOG_RATE_FIRST -> {
                    getReviewInfo()
                    viewState.onQuestionDialog(
                        context.getString(R.string.dialogs_question_rate_second_info), TAG_DIALOG_RATE_SECOND,
                        context.getString(R.string.dialogs_question_accept_sure),
                        context.getString(R.string.dialogs_question_accept_no_thanks), null
                    )
                }

                TAG_DIALOG_RATE_SECOND -> {
                    viewState.onDialogClose()
                    reviewInfo?.let {
                        viewState.onShowInAppReview(it)
                    } ?: run {
                        viewState.onShowPlayMarket(BuildConfig.RELEASE_ID)
                    }
                }

                TAG_DIALOG_RATE_FEEDBACK -> {
                    if (value != null) {
                        viewState.onShowEmailClientTemplate(value)
                    }
                    viewState.onDialogClose()
                }
            }
        }
    }

    fun onCancelClick(tag: String?) {
        tag?.let {
            when (tag) {
                TAG_DIALOG_RATE_FIRST -> {
                    preferenceTool.isRateOn = false
                    viewState.onDialogClose()
                    viewState.onShowEditMultilineDialog(
                        context.getString(R.string.dialogs_edit_feedback_rate_title),
                        context.getString(R.string.dialogs_edit_feedback_rate_hint),
                        context.getString(R.string.dialogs_edit_feedback_rate_accept),
                        context.getString(R.string.dialogs_question_accept_no_thanks), TAG_DIALOG_RATE_FEEDBACK
                    )
                }

                TAG_DIALOG_RATE_SECOND -> {
                    preferenceTool.isRateOn = false
                    viewState.onDialogClose()
                }
            }
        }
    }

    fun clear() {
        accountPreferences.onlineAccountId = null
    }

    fun onRateOff() {
        preferenceTool.isRateOn = false
    }

    fun checkOnBoardingShowed() {
        if (!preferenceTool.onBoarding) {
            viewState.onShowOnBoarding()
        }
    }

    fun openDeeplink(uri: Uri) {
        val data = Json.decodeFromString<OpenDataModel>(CryptUtils.decodeUri(uri.query))
        val hasToken = data.share.isNotEmpty()
        val account = context.accountOnline
        val isAccountOnline = account?.portal?.urlWithScheme == data.portal &&
                account?.login == data.email

        presenterScope.launch(Dispatchers.Default) {
            if (isAccountOnline || hasToken) {
                openFile(data)
                return@launch
            }

            when (val result = accountRepository.checkLoginWithEmailAndPortal(email = data.email.orEmpty(), portalUrl = data.portal)) {
                is CheckLoginResult.Success -> {
                    viewState.restartActivity(deeplink = uri)
                }
                is CheckLoginResult.NeedLogin -> {
                    signInAndOpenDeeplink(data, uri)
                }
                is CheckLoginResult.Error -> fetchError(result.exception)
                else -> Unit
            }
        }
    }

    private suspend fun signInAndOpenDeeplink(data: OpenDataModel, uri: Uri) {
        val portalUri = data.portal?.toUri() ?: return
        val portalScheme = portalUri.scheme?.let { Scheme.valueOf("$it://") }
            ?: data.originalUrl?.toUri()?.scheme?.let { Scheme.valueOf("$it://") }
            ?: Scheme.Https

        App.getApp().refreshLoginComponent(
            CloudPortal(url = portalUri.host.takeIf { it.isNullOrEmpty() } ?: portalUri.toString(), scheme = portalScheme)
        )

        App.getApp().loginComponent.cloudLoginRepository
            .checkPortal(portalUri.host.orEmpty(), portalScheme)
            .collect { result ->
                withContext(Dispatchers.Main) {
                    when (result) {
                        is PortalResult.Error -> {
                            viewState.onError(context.getString(R.string.errors_unknown_error))
                        }

                        is PortalResult.Success -> {
                            App.getApp().refreshLoginComponent(result.cloudPortal)
                            viewState.signInAndOpenDeeplink(
                                portal = data.portal,
                                email = data.email.orEmpty(),
                                uri = uri
                            )
                        }

                        else -> Unit
                    }
                }
            }
    }

    private suspend fun openFile(data: OpenDataModel) {
        presenterScope.launch {
//            if (data.folder != null) return@launch

            recentDataSource.insertOrUpdate(Recent(
                fileId = data.file?.id.orEmpty(),
                name = data.file?.title.orEmpty(),
                source = data.portal,
                token = data.share
            ))
        }

        context.cloudFileProvider
            .openDeeplink(
                portal = data.portal.orEmpty(),
                token = data.share,
                login = data.email.orEmpty(),
                id = data.file?.id.orEmpty(),
                extension = data.file?.extension.orEmpty(),
                title = data.file?.title.orEmpty()
            )
            .collect { result ->
                when (result) {
                    is NetworkResult.Error -> fetchError(result.exception)
                    is NetworkResult.Success<FileOpenResult> -> {
                        handleFileOpenResult(result.data)
                    }
                    else -> Unit
                }
            }
    }

    private fun handleFileOpenResult(result: FileOpenResult) {
        when (result) {
            is FileOpenResult.OpenDocumentServer -> with(result) {
                viewState.onDialogClose()
                viewState.showEditors(
                    data = info,
                    extension = cloudFile.fileExst,
                    access = cloudFile.access,
                    editType = editType,
                    onResultListener = null
                )
            }

            is FileOpenResult.OpenLocally -> with(result) {
                viewState.onDialogClose()
                viewState.showEditors(
                    uri = file.toUri(),
                    extension = file.extension,
                    editType = editType,
                    access = access,
                    onResultListener = null
                )
            }

            else -> Unit
        }
    }

    fun setCheckVPN(checked: Boolean) {
        preferenceTool.isVpnChecked = checked
    }

}